### PR TITLE
BAN-1470: Implement Pro UI (bonding curve) on Lend Page

### PR DESCRIPTION
**Pull request by iamsphere**
## Description

Implement Pro UI (bonding curve) on Lend Page

## Issue link

https://linear.app/banx-gg/issue/BAN-1470/fe-banx-implement-pro-ui-bonding-curve-on-lend-page

## Vercel preview

https://banx-ui-git-feature-ban-1470-frakt.vercel.app/

### CODE REVIEW

**Review by sablevsky — CHANGES_REQUESTED** [review]

File: src/transactions/loans/makeRepayPartialLoanAction.ts [hunk 1]
```diff
@@ -22,6 +24,8 @@ export type MakeRepayPartialLoanAction = MakeActionFn<
   MakeRepayPartialActionResult
 >
 
+type OptimisticResult = BondAndTransactionOptimistic & { oldBondOffer: BondOfferV2 }
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/transactions/loans/makeRepayPartialLoanAction.ts]: Спулился

File: src/store/offers/useSyntheticOffers.ts [hunk 2]
```diff
@@ -9,8 +9,10 @@ export interface SyntheticOffer {
   publicKey: string //? PUBKEY_PLACEHOLDER for offers to create
   loanValue: number
   loansAmount: number
+  deltaValue?: number //? exist only in pro mode
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/store/offers/useSyntheticOffers.ts]: [Add delta and mathCounter as required field, set 0 as default](https://github.com/frakt-solana/banx-ui/pull/280/commits/4ee397272a0804ac8abaac6e6a966bb06fc0f10e)
Сделал

File: src/store/offers/useSyntheticOffers.ts [hunk 3]
```diff
@@ -9,8 +9,10 @@ export interface SyntheticOffer {
...
   assetReceiver: string
   marketPubkey: string
+  mathCounter?: number
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/store/offers/useSyntheticOffers.ts]: [Add delta and mathCounter as required field, set 0 as default](https://github.com/frakt-solana/banx-ui/pull/280/commits/4ee397272a0804ac8abaac6e6a966bb06fc0f10e)

File: src/store/offers/useSyntheticOffers.ts [hunk 4]
```diff
@@ -61,10 +63,18 @@ export const createEmptySyntheticOffer: CreateEmptySyntheticOffer = ({
   loansAmount: 0,
   assetReceiver: walletPubkey,
   marketPubkey,
+  mathCounter: 0,
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/store/offers/useSyntheticOffers.ts:66]: [Add delta and mathCounter as required field, set 0 as default](https://github.com/frakt-solana/banx-ui/pull/280/commits/4ee397272a0804ac8abaac6e6a966bb06fc0f10e)

File: src/store/offers/useSyntheticOffers.ts [hunk 5]
```diff
@@ -76,5 +86,6 @@ export const convertToSynthetic = (offer: Offer, isEdit = false): SyntheticOffer
     loanValue,
     assetReceiver,
     marketPubkey: hadoMarket,
+    mathCounter,
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/store/offers/useSyntheticOffers.ts:91]: [Add delta and mathCounter as required field, set 0 as default](https://github.com/frakt-solana/banx-ui/pull/280/commits/4ee397272a0804ac8abaac6e6a966bb06fc0f10e)

File: src/store/offers/useSyntheticOffers.ts [hunk 6]
```diff
@@ -61,10 +63,18 @@ export const createEmptySyntheticOffer: CreateEmptySyntheticOffer = ({
...
 export const convertToSynthetic = (offer: Offer, isEdit = false): SyntheticOffer => {
-  const { fundsSolOrTokenBalance, currentSpotPrice, publicKey, assetReceiver, hadoMarket } = offer
+  const {
+    fundsSolOrTokenBalance,
+    currentSpotPrice,
+    publicKey,
+    assetReceiver,
+    hadoMarket,
+    mathCounter,
+  } = offer
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/store/offers/useSyntheticOffers.ts]: [Add delta and mathCounter as required field, set 0 as default](https://github.com/frakt-solana/banx-ui/pull/280/commits/4ee397272a0804ac8abaac6e6a966bb06fc0f10e)

File: src/transactions/bonds/makeCreateBondingOfferAction.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+} from 'fbonds-core/lib/fbond-protocol/functions/perpetual'
+import { BondingCurveType } from 'fbonds-core/lib/fbond-protocol/types'
+import { MakeActionFn } from 'solana-transactions-executor'
+
+import { BONDS } from '@banx/constants'
+import { sendTxnPlaceHolder } from '@banx/utils'
+
+export type MakeCreateBondingOfferActionParams = {
+  marketPubkey: string
+  loanValue: number //? value in sol
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/transactions/bonds/makeCreateBondingOfferAction.ts]: https://github.com/frakt-solana/banx-ui/pull/280/commits/49321c8b2e52fd434cabac8f5af38ff884e5ea0e
Помел тут

File: src/transactions/bonds/makeUpdateBondingOfferAction.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+} from 'fbonds-core/lib/fbond-protocol/functions/perpetual'
+import { BondOfferV2 } from 'fbonds-core/lib/fbond-protocol/types'
+import { MakeActionFn } from 'solana-transactions-executor'
+
+import { Offer } from '@banx/api/core'
+import { BONDS } from '@banx/constants'
+import { sendTxnPlaceHolder } from '@banx/utils'
+
+export type MakeUpdateBondingOfferActionParams = {
+  loanValue: number //? value in sol
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/transactions/bonds/makeUpdateBondingOfferAction.ts:14]: https://github.com/frakt-solana/banx-ui/pull/280/commits/49321c8b2e52fd434cabac8f5af38ff884e5ea0e
Поменял тут

File: src/pages/LendPage/components/PlaceOfferTab/PlaceProOffer/components.tsx [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+          />
+          <StatInfo label="Reserve" value={MOCK_RESERVE_VALUE} tooltipText="Reserve" />
+          <StatInfo label="Accrued interest" value={MOCK_ACCRUED_INTEREST} />
+        </div>
+      )}
+    </div>
+  )
+}
+
+type CalculateWeightedWeeklyInterest = (offerSize: number, marketApr: number) => number
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/PlaceProOffer/components.tsx]: https://github.com/frakt-solana/banx-ui/pull/280/commits/49321c8b2e52fd434cabac8f5af38ff884e5ea0e
Поправил

File: src/pages/LendPage/components/PlaceOfferTab/PlaceLiteOffer/PlaceLiteOffer.tsx [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        offerSize={offerSize}
+        marketAPR={marketApr}
+        loanToValuePercent={loanToValuePercent}
+        isEditMode={isEditMode}
+      />
+      <OfferActionButtons
+        isEditMode={isEditMode}
+        disableUpdateOffer={disableUpdateOffer}
+        disablePlaceOffer={disablePlaceOffer}
+        {...offerTransactions}
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/PlaceLiteOffer/PlaceLiteOffer.tsx]: [Remove spread for better readability](https://github.com/frakt-solana/banx-ui/pull/280/commits/e1d9f0681aa8ff4a75573d859fc351fe46070041)

File: src/pages/LendPage/components/PlaceOfferTab/PlaceProOffer/PlaceProOffer.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+      <div className={styles.messageContainer}>
+        {showDepositError && <InputErrorMessage message="Not enough SOL" />}
+        {showBorrowerMessage && <BorrowerMessage loanValue={loanValue} />}
+      </div>
+      <OfferSummary isEditMode={isEditMode} offerSize={offerSize} marketApr={marketApr} />
+      <OfferActionButtons
+        isEditMode={isEditMode}
+        disableUpdateOffer={disableUpdateOffer}
+        disablePlaceOffer={disablePlaceOffer}
+        {...offerTransactions}
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/PlaceProOffer/PlaceProOffer.tsx]: [Remove spread for better readability](https://github.com/frakt-solana/banx-ui/pull/280/commits/e1d9f0681aa8ff4a75573d859fc351fe46070041)

File: src/pages/LendPage/components/PlaceOfferTab/hooks/usePlaceOffer.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  onChangeOfferMode: (value: OfferMode) => void
+  setSyntheticOffer: (offer: SyntheticOffer) => void
+}
+
+type UsePlaceOffer = (props: OrderBookMarketParams) => OfferParams & { offerMode: OfferMode }
+
+export const usePlaceOffer: UsePlaceOffer = (props) => {
+  const { publicKey: walletPubkey } = useWallet()
+
+  const { offerMode, marketPubkey, onChangeOfferMode, setOfferPubkey, offerPubkey } = props || {}
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/hooks/usePlaceOffer.ts]: [Remove unnecessary nullish coalescing in prop destructuring](https://github.com/frakt-solana/banx-ui/pull/280/commits/d7409ab139686777b237df27366f6ad3c78b973e)

File: src/pages/LendPage/components/PlaceOfferTab/hooks/usePlaceOffer.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+  optimisticOffer: Offer | undefined
+  syntheticOffer: SyntheticOffer
+
+  exitEditMode: () => void
+  updateOrAddOffer: (offer: Offer) => void
+  onChangeOfferMode: (value: OfferMode) => void
+  setSyntheticOffer: (offer: SyntheticOffer) => void
+}
+
+type UsePlaceOffer = (props: OrderBookMarketParams) => OfferParams & { offerMode: OfferMode }
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/hooks/usePlaceOffer.ts]: d7409ab139686777b237df27366f6ad3c78b973e

File: src/pages/LendPage/components/PlaceOfferTab/PlaceLiteOffer/components.tsx [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+  loanToValuePercent,
+  isEditMode,
+}) => {
+  const weeklyAprPercentage = marketAPR / 100 / WEEKS_IN_YEAR
+  const estimatedInterest = (offerSize * weeklyAprPercentage) / 100
+
+  const colorLTV = getColorByPercent(loanToValuePercent, HealthColorIncreasing)
+
+  const displayEstimatedInterest = estimatedInterest ? formatDecimal(estimatedInterest) : 0
+  const displayOfferSize = offerSize ? formatDecimal(offerSize) : 0
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/PlaceLiteOffer/components.tsx]: [convert variables to one type](https://github.com/frakt-solana/banx-ui/pull/280/commits/5da42324c22f0627389a161d892474ac2a385eb5)

File: src/pages/LendPage/components/PlaceOfferTab/PlaceLiteOffer/components.tsx [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  marketAPR,
+  loanToValuePercent,
+  isEditMode,
+}) => {
+  const weeklyAprPercentage = marketAPR / 100 / WEEKS_IN_YEAR
+  const estimatedInterest = (offerSize * weeklyAprPercentage) / 100
+
+  const colorLTV = getColorByPercent(loanToValuePercent, HealthColorIncreasing)
+
+  const displayEstimatedInterest = estimatedInterest ? formatDecimal(estimatedInterest) : 0
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/PlaceLiteOffer/components.tsx]: [convert variables to one type](https://github.com/frakt-solana/banx-ui/pull/280/commits/5da42324c22f0627389a161d892474ac2a385eb5)

File: src/pages/LendPage/components/PlaceOfferTab/PlaceLiteOffer/helpers.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+const TRANSACTION_FEE_IN_SOL = 0.01
+
+export const calculateBestLoanValue = (solanaBalance: number, bestOffer: number) => {
+  const balanceAfterDeductingFee = solanaBalance - TRANSACTION_FEE_IN_SOL
+  const maxLoanValue = Math.max(balanceAfterDeductingFee, 0)
+
+  const bestOfferInSol = bestOffer / 1e9 || 0
+
+  const defaultLoanValue = Math.min(maxLoanValue, bestOfferInSol) || 0
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/PlaceLiteOffer/helpers.ts:11]: [Remove unnecessary default value in bestOfferInSol calculation](https://github.com/frakt-solana/banx-ui/pull/280/commits/cabc39807f4c4e55267e74b87bc505d124dcd2df)

File: src/pages/LendPage/components/PlaceOfferTab/PlaceLiteOffer/PlaceLiteOffer.tsx [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+  return (
+    <>
+      <div className={styles.fields}>
+        <NumericInputField
+          label="Offer"
+          value={loanValue}
+          onChange={onLoanValueChange}
+          className={styles.numericField}
+          disabled={!connected}
+          hasError
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/PlaceLiteOffer/PlaceLiteOffer.tsx]: Убрал в этом коммите
e1d9f0681aa8ff4a75573d859fc351fe46070041

File: src/pages/LendPage/components/PlaceOfferTab/PlaceLiteOffer/helpers.ts [hunk 18]
```diff
@@ -0,0 +1,22 @@
+import { MarketPreview } from '@banx/api/core'
+
+const TRANSACTION_FEE_IN_SOL = 0.01
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/PlaceLiteOffer/helpers.ts]: Можно попробовать и меньше, но 0.01 запревентит любые кейсы с "not enough sol"

File: src/pages/LendPage/components/PlaceOfferTab/hooks/useOfferTransactions.ts [hunk 19]
```diff
@@ -16,27 +17,70 @@ export const useOfferTransactions = ({
...
+        defaultTxnErrorHandler(error, {
+          additionalData: txnParam,
+          walletPubkey: wallet?.publicKey?.toBase58(),
+          transactionName: 'UpdateBondingOffer',
+        })
+      })
+      .execute()
+  }
 
   const onCreateOffer = async () => {
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/hooks/useOfferTransactions.ts]: [Remove old onCreateOffer, onUpdateOffer functions](https://github.com/frakt-solana/banx-ui/pull/280/commits/502b08c409bab1d4902c4cca6f2658647a6159b3)

File: src/pages/LendPage/components/PlaceOfferTab/hooks/useOfferTransactions.ts [hunk 20]
```diff
@@ -63,6 +104,28 @@ export const useOfferTransactions = ({
       .execute()
   }
 
+  const onUpdateOffer = async () => {
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/hooks/useOfferTransactions.ts]: [Remove old onCreateOffer, onUpdateOffer functions](https://github.com/frakt-solana/banx-ui/pull/280/commits/502b08c409bab1d4902c4cca6f2658647a6159b3)

File: src/pages/LendPage/components/PlaceOfferTab/hooks/useOfferTransactions.ts [hunk 21]
```diff
@@ -90,31 +150,44 @@ export const useOfferTransactions = ({
...
+    onCreateBondingOffer,
+    onUpdateBondingOffer,
+    onClaimOfferInterest,
+    onRemoveOffer,
+    onCreateOffer,
+    onUpdateOffer,
+  }
+}
+
+const showSuccessSnackbar = ({ message, txnHash }: { message: string; txnHash: string }) => {
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/hooks/useOfferTransactions.ts]: Убрал вообще эту функцию
[Remove old onCreateOffer, onUpdateOffer functions](https://github.com/frakt-solana/banx-ui/pull/280/commits/502b08c409bab1d4902c4cca6f2658647a6159b3)

File: src/pages/LendPage/components/PlaceOfferTab/PlaceOfferTab.tsx [hunk 22]
```diff
@@ -1,70 +1,28 @@
...
-    offerTransactions,
-    offerSize,
-    marketApr,
-    showDepositError,
-    showBorrowerMessage,
-    disableUpdateOffer,
-    disablePlaceOffer,
-    loanToValuePercent,
-  } = usePlaceOfferTab({ ...props })
+  const isEditMode = !!offerPubkey && offerPubkey !== PUBKEY_PLACEHOLDER
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/PlaceOfferTab/PlaceOfferTab.tsx:18]: В 2х местах используется, думаю пока можно не выносить

File: src/pages/LendPage/components/OrderBook/OrderBook.module.less [hunk 23]
```diff
@@ -1,160 +1,144 @@
...
-.content {
+.orderBook {
   display: flex;
   flex-direction: column;
-  gap: 24px;
-  position: relative;
-  height: 100%;
+  margin: 16px 24px 0;
+  height: calc(100% - 72px - 16px); //? 72px => accrued interest container, 16px => labels container
+  min-width: 296px;
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/pages/LendPage/components/OrderBook/OrderBook.module.less]: [Remove css calculations in OrderBook.module.less](https://github.com/frakt-solana/banx-ui/pull/280/commits/d32327d8dd6235322c72425aed829c9740f380bd)

File: src/utils/common.ts [hunk 24]
```diff
@@ -34,7 +34,9 @@ export const getDecimalPlaces = (value: number | string) => {
   return numericValue < DECIMAL_THRESHOLD ? THREE_DECIMAL_PLACES : TWO_DECIMAL_PLACES
 }
 
-export const formatDecimal = (value: number) => {
+export const formatDecimal = (value: number | string) => {
+  if (!value) return '0'
```
[the remark this answers is not in the document]
Comment (iamsphere) [reply src/utils/common.ts]: [Modify formatDecimal, getDecimalPlaces functions](https://github.com/frakt-solana/banx-ui/pull/280/commits/9f3b192b965473920752eed0090c58b292a18515)

File: src/pages/LendPage/components/PlaceOfferTab/PlaceOfferTab.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (sablevsky) [reply src/pages/LendPage/components/PlaceOfferTab/PlaceOfferTab.tsx:18]: Оке

**Review by sablevsky — APPROVED** [review]

**Pull request merged**